File: NaryTree/Python/n-ary_tree_postorder_traversal.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: Optional[int] = None, children: Optional[List['Node']] = None):
        self.val = val
        self.children = children
"""

class Solution:
    def postorder(self, root: 'Node') -> list[int]:
        if not root: return []
        stack, ans = [root], []
        while stack:
            cur = stack.pop()
            ans.append(cur.val)
            stack.extend(cur.children)
        return ans[::-1]

    # recursive
    def postorder(self, root: 'Node') -> list[int]:
        if not root: return []
        ans = []
        for child in root.children:
            ans.extend(self.postorder(child))
        ans.append(root.val)
        return ans

File: NaryTree/JS/n_ary_tree_postorder_traversal.js
/**
 * // Definition for a _Node.
 * function _Node(val,children) {
 *    this.val = val;
 *    this.children = children;
 * };
 */

/**
 * @param {_Node|null} root
 * @return {number[]}
 */
var postorder = function(root) {
    if (!root) return [];
    let ans = [];
    for (const child of root.children) {
        const cur = postorder(child);
        ans.push(...cur);
    }
    ans.push(root.val);
    return ans;
};

//iterative
var postorder = function(root) {
    if (!root) return [];
    let ans = [];
    let stack = [root];
    while (stack.length) {
        const cur = stack.pop();
        ans.push(cur.val);
        stack.push(...cur.children);
    }
    return ans.reverse();
};